sw/device/tests/crypto/cryptotest/firmware/kmac_sca.c

```diff
@@ -512,27 +521,45 @@ status_t handle_kmac_sca_single_absorb(ujson_t *uj) {
   }
 
   // Ungate the capture trigger signal and then start the operation.
+  dif_kmac_operation_state_t operation_state;
   sca_set_trigger_high();
-  if (sha3_ujson_absorb(uj_msg.msg, uj_msg.msg_length) != kmacScaOk) {
-    return ABORTED();
+  if (fpga_mode) {
+    if (sha3_ujson_absorb(uj_msg.msg, uj_msg.msg_length) != kmacScaOk) {
+      return ABORTED();
+    }
+  } else {
+    UJSON_CHECK_DIF_OK(dif_kmac_mode_kmac_start(
+        &kmac, &operation_state, kDifKmacModeKmacLen128, kDigestLength,
+        &kmac_key, NULL));
+    UJSON_CHECK_DIF_OK(dif_kmac_absorb(&kmac, &operation_state, uj_msg.msg,
+                                       uj_msg.msg_length, NULL));
   }
   sca_set_trigger_low();
 
-  // Check KMAC has finsihed processing the message.
-  kmac_msg_done();
-
-  // Read the digest and send it to the host for verification.
   uint32_t out[kDigestLength];
-  if (kmac_get_digest(out, kDigestLength) != kmacScaOk) {
-    return ABORTED();
+  if (fpga_mode) {
+    // Check KMAC has finished processing the message.
+    kmac_msg_done();
+
+    // Read the digest and send it to the host for verification.
+    if (kmac_get_digest(out, kDigestLength) != kmacScaOk) {
+      return ABORTED();
+    }
+  } else {
+    UJSON_CHECK_DIF_OK(
+        dif_kmac_squeeze(&kmac, &operation_state, out, kDigestLength, NULL));
   }
```

Comment:
Also, we've just discussed to investigate whether we can split the derived functions a bit further to do the following sequence:
1. Setup
2. Start CMD -  this should already trigger the prefix and key hashing ops, put Ibex to sleep
3. Write message FIFO
4. Process CMD, put Ibex to sleep
5. Read out digest

If this works, it would allow to not overlap the message FIFO writes with the hashing of the prefix and the key. Before doing a full implementation of this. It's probably worth to do some quick trials to check whether this works indeed.

Reply:
Thanks for your comment @vogelpi.

I've totally missed that the PROCESS command is issued in `dif_kmac_squeeze()`, so we would miss operations.

Now, in `fpga_mode=false`, I am doing the following in the `sha3_ujson_absorb()` function, which is called when the capture trigger is high:
 - Configure mode/key/...
 - Send START command, put Ibex into sleep.
 - Write message FIFO.
 - Send PROCESS command, put Ibex into sleep.
The digest is retrieved afterwards.

I've tested the changes using your bitstream from your ot-sca-cw310 [branch](https://github.com/vogelpi/opentitan/commit/7e52860a57b34dacdf3de71f45d754cb1cacb808). As expected, when I set `fpga_mode=true`, capture does not work, i.e., I retrieve wrong data from the device. In the `fpga_mode=false`, I retrieve the correct data with your bitstream and the new device binary.

Does this look correct for you?

The only thing missing for now is the correct number of sleep cycles for the `sca_call_and_sleep()` calls in `fpga_mode=false`.